### PR TITLE
network: add RouteMetric= in [DHCPv6] section

### LINKED ISSUE #15295: IPv6 and RouteMetric
**systemd version the issue has been seen with**
> systemd 245 (245.3-2-arch)
> +PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

**Used distribution**
 > ArchLinux

**Expected behaviour you didn't see**
 > Routing table entry with only the correct RouteMetric set

**Unexpected behaviour you saw**
 > Dupplicated routing table entry once with metric 1024 and once with the provided one.

**Steps to reproduce the problem**
Use a network unit like:
```
[Match]
Name=wlp2s0

[Network]
DHCPServer=no
DHCP=yes
DNSOverTLS=opportunistic
DNSSEC=allow-downgrade
DNS=2606:4700:4700::1111 2001:4860:4860::8888 1.1.1.1 8.8.8.8
DNSDefaultRoute=yes
IPv6AcceptRA=yes
IPv6DuplicateAddressDetection=1
IPv6PrivacyExtensions=kernel
LLMNR=no
MulticastDNS=no

[DHCP]
Anonymize=yes
UseDNS=no
UseDomains=route
UseHostname=no
RouteMetric=2000
```
Than the routing table looks like:
```
Destination                    Next Hop                   Flag Met Ref Use If
::/0                           fe80::feec:daff:fe1c:76b2  UGe  1024 6     0 enp0s25
::/0                           fe80::feec:daff:fe1c:76b2  UGe  1024 5     0 wlp2s0
::/0                           fe80::feec:daff:fe1c:76b2  UGe  2000 1     0 wlp2s0
```
obviously the 2nd line should not be present.

Note: The documentation at https://www.freedesktop.org/software/systemd/man/systemd.network.html also only lists the option `RouteMetric` for the section `DHCPv4` instead of also for `DHCP`.

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #15295.

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -1734,6 +1734,14 @@
           </listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term><varname>RouteMetric=</varname></term>
+          <listitem>
+            <para>Set the routing metric for routes specified by the
+            DHCP server.</para>
```
Comment (keszybz) [inline man/systemd.network.xml]: Shouldn't this say that 1024 will be used by default?

File: src/network/networkd-dhcp-common.c [hunk 2]
```diff
@@ -63,6 +63,40 @@ int config_parse_dhcp(
...
+
+        r = safe_atou32(rvalue, &metric);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to parse RouteMetric=%s, ignoring assignment: %m", rvalue);
+                return 0;
+        }
+
+        network->dhcp_route_metric = r;
+        network->dhcp6_route_metric = r;
```
Comment (keszybz) [inline src/network/networkd-dhcp-common.c]: Hmm, that means that if specify
```
[DHCPv6]
RouteMetric=1
[DHCP]
RouteMetric=2
```
I'll get 2 for dhcp6_route_metric. Shouldn't this only be set if the other setting is not used?

Comment (yuwata) [reply src/network/networkd-dhcp-common.c]: `[DHCP]` section is now deprecated. So, I think it is not necessary to consider the case that `[DHCP]` and `[DHCPv6]` sections are simultaneously used.

File: man/systemd.network.xml (hunk 1 above)
Comment (yuwata) [reply man/systemd.network.xml]: OK. Updated.

File: src/network/networkd-dhcp-common.c (hunk 2 above)
Comment (ssahani) [reply src/network/networkd-dhcp-common.c]: Why we are setting the both ?

Comment (keszybz) [reply src/network/networkd-dhcp-common.c]: > Why we are setting the both ?

For backwards compatibility. See the commit message.

> `[DHCP]` section is now deprecated. So, I think it is not necessary to consider the case that `[DHCP]` and `[DHCPv6]` sections are simultaneously used.

Actually, I think people might reasonably set both, to support newer and older systemd simultaneously.

Comment (ssahani) [reply src/network/networkd-dhcp-common.c]: DHCPv6 route metric never supported and `[DHCP]` section is deprecated.

Comment (keszybz) [reply src/network/networkd-dhcp-common.c]: "deprecated" means "supported, but not encouraged". So if something is deprecated, we still need handle things correctly.

Comment (yuwata) [reply src/network/networkd-dhcp-common.c]: OK. I changed my mind. Now `[DHCPv4]` or `[DHCPv6]` sections take precedence. PTAL.

**Comment by ssahani** [comment]
LGTM.

**Comment by keszybz** [comment]
LGTM.

**Pull request merged**